### DISCUSSION TITLE
Implementing rs*_get_EVP_MD() without EVP_MD_meth_dup()

### DISCUSSION THREAD

**Discussion by botovq** [Yubico/libfido2#794] [General]
We removed `EVP_MD_meth_dup()` from LibreSSL 3.9.0 and later because it is part of a dangerous API that is essentially unused. In OpenBSD, I fixed this with [this patch][1] that simplifies and improves the code quite a bit in my opinion.

I was going to send this diff  in a PR earlier today, but, unfortunately, the combination of `-Werror` with `-Wcast-qual` results in a compilation error.

I see four solutions for this - assuming you want to keep both these compiler flags for all builds:

1. Add compiler pragmas to the two functions where we need to cast away const, so `-Wcast-qual` is happy and use the fix I landed in OpenBSD.
2. It is possible to cast away const by passing through a `uintptr_t`:
   ```C
    static EVP_MD *
    rs256_get_EVP_MD(void)
    {
            const EVP_MD *md = EVP_sha256();
            return (EVP_MD *)(uintptr_t)md;
    }
    ```
3. We include `EVP_MD_meth_dup()` and `EVP_MD_meth_free()` in the stable version of LibreSSL 3.9.
4. Do some hacking with the precompiler using a macro that is defined to empty in the OpenSSL 3 path and to `const` in the LibreSSL/legacy OpenSSL path.

If we ever add `EVP_MD_fetch()` to LibreSSL, it will basically be a version of 1. @davidben from BoringSSL is of the same opinion.

Solution 4 will be quite ugly, I would love to avoid 3 if at all possible and I don't particularly like solution 2, which leaves me with a strong preference for solution 1.

Cc @djmdjm

[1]: https://github.com/openbsd/src/commit/aeab8fe0aafabfb19384dbffab319b32735102a8

**Comment by davidben** [comment]
(1) is also best matches how one typical applications use OpenSSL 1.1.1, and is more performant since you avoid making a dummy copy of the EVP_MD when you didn't need to.

TBH you probably could just use that in 3.x too. While OpenSSL does say fetch is more performant, AIUI that's only if the application saves the EVP_MD to work around the various perf regressions in 3.x.

**Comment by martelletto** [comment]
I agree solution 1 is the best way to go.

**Comment by LDVG** [comment]
+1, that seems like the sensible way forward. As @davidben suggests, that could also let us drop the EVP_MD_fetch() call.
